### DISCUSSION TITLE
planning an `iroh://` URI scheme

### DISCUSSION THREAD

**Discussion by b5** [n0-computer/iroh#1928] [Ideas]
We've had a [question on discord](https://discord.com/channels/1161119546170687619/1161119546644627528/1192680611832418385) about implementing an `iroh://` scheme:

> Has there been any discussion of what an iroh URI would look like? I want to write a small URI rewriter that takes any requests for iroh:// or something and redirects them to the iroh-examples gateway. Initially I think I'll just take iroh:// and replace that with http://localhost:8080/ so that I can handle these routes: https://github.com/n0-computer/iroh-examples/blob/67e58859c74941be595ef9526ccbb5f9f721cbd7/iroh-gateway/src/main.rs#L509-L513 like iroh://blob/bafkr4ibwueokpaxz3x6fd2tvsdo3uqixzimulg536abicpqmqtt3itaodt, but maybe there's been more of a discussion of this somewhere.

This is a spec-level thing, so let's kick off a chat about the scope of what a scheme like `iroh://` should do, and ideally draft up a minimum surface area to start with.

**Comment by expenses** [comment]
Hi, so I brought this up in the discord chat. One use case I'm looking into is writing a plugin for [OpenUSD](https://openusd.org/release/index.html) that allows for [resolving assets](https://openusd.org/release/api/ar_page_front.html) distributed across a network. Initially just having a URI scheme that allows for referencing content-addressed blobs would be fine, and something like `iroh://blob/<hash>` (similar to ipfs) would work fine.

In the longer term it'd be nice to have a way to reference a document to subscribe to and get a persistent stream of updates from. `iroh://watch/<doc>` or something like that might work? 

One requirement of the OpenUSD asset resolver API mentioned above is that filetypes need to be determined before the asset bytes are returned. With http and most other URIs this is easily done because you can use the file extension 99% of the time. For something like an `iroh://` address that lacks an extension, you could just read the first few bytes and determine the mime type and that should be good enough, but for less flexible systems it might be valuable to allow for adding on an extension after the hash like `iroh://blob/<hash>.<ext>`.

**Reply by dpc** [reply]
iroh blob is a protocol, so probably `iroh+blob://<node_id>/<whateverelse` is the right format.

**Reply by matheus23** [reply]
This is now correct. For completion's sake I want to point out that at the time that expenses wrote this comment, "blobs" was part of what we called "iroh" (as was iroh-gossip and iroh-docs).

**Comment by dpc** [comment]
https://github.com/n0-computer/iroh/discussions/3140 - previous discussion
